README.md
[![Build Status](https://travis-ci.org/juicecube/mlz-axios.svg?branch=master)](https://travis-ci.org/juicecube/mlz-axios)
[![codecov](https://codecov.io/gh/juicecube/mlz-axios/branch/master/graph/badge.svg)](https://codecov.io/gh/juicecube/mlz-axios)

## 介绍

对 axios 做了二次封装，统一 axios 的配置

！！！！！！请求/响应拦截放在实例化之前！！！！！！
！！！！！！请求/响应拦截放在实例化之前！！！！！！
！！！！！！请求/响应拦截放在实例化之前！！！！！！

## 安装

使用 npm

```
npm install @mlz/axios
```

使用 yarn

```
yarn add @mlz/axios
```

## 默认 axios 配置

```js
timeout: 5000;
withCredentials: true;
validateStatus: status => status >= 200 && status < 599;
```

## 实例化

实例化可以传入 baseUrl 和默认配置

```js
import Http from 'mlz-axios'
const httpIns = new Http(baseUrl:string, config:AxiosRequestConfig)
```

## 静态方法
### 获取对应的实例
```js
getInstances(key:string)
```
示例:
```js
import Http from "mlz-axios";
Http.getInstances("https://www.example.com");
```

### 全局设置token与tokenType
```js
setGlobalTokenConfig({
  authorizationTypeKey?: string;
  authorizationTokenKey?: string;
  authorizationTypeValue?: number;
  authorizationTokenValue?: string;
})
```
示例：
```js
import Http from "mlz-axios";
Http.setGlobalTokenConfig({
  authorizationTypeKey: 'authorization_type',
  authorizationTokenKey: 'Authorization',
  authorizationTypeValue: 1
  authorizationTokenValue: 'token'
});
```
设置请求拦截器
```js
setReqInterceptor (resolve?:(config:AxiosRequestConfig) => AxiosRequestConfig, reject?:(error:any) => void, url?:string)
```
设置全局请求拦截器
```js
import Http from "mlz-axios";
Http.setReqInterceptor(
  config => {
    console.log(config);
  },
  err => {
    console.log(err);
  }
);
```

设置单个实例请求拦截器

```js
Http.setReqInterceptor(
  config => {
    console.log("请求前config", config);
    return config;
  },
  err => {
    console.log("请求前err", err);
    return Promise.reject(err);
  },
  "https://xxx.xxx.com"
);
```
### 设置响应拦截器
```js
setResInterceptor (resolve?:(res:AxiosResponse) => AxiosResponse, reject?:(error) => any, url?:string)
```

设置全局响应拦截器

```js
Http.setResInterceptor(
  res => {
    console.log("请求后res", res);
    return res;
  },
  err => {
    console.log("请求后err", err);
    return Promise.reject(err);
  }
);
```

设置单个实例响应拦截器

```js
Http.setResInterceptor(
  res => {
    console.log("请求后res", res);
    return res;
  },
  err => {
    console.log("请求后err", err);
    return Promise.reject(err);
  },
  "https://xxx.xxx.com"
);
```

## 实例方法
设置实例的 authorizationTokenToken 和 authorizationType
```js
setInstanceTokenConfig({
  authorizationTypeKey?: string;
  authorizationTokenKey?: string;
  authorizationTypeValue?: number;
  authorizationTokenValue?: string;
})
```
示例：
```js
import Http from "mlz-axios";
const httpIns = new Http('https://xxx.xxx.com')
httpIns.setInstanceTokenConfig({
  authorizationTypeKey: 'authorization_type',
  authorizationTokenKey: 'Authorization',
  authorizationTypeValue: 1
  authorizationTokenValue: 'token'
});

```

### get(url[, config])

```js
import Http from "mlz-axios";
const httpIns = new Http("https://www.example.com");
httpIns
  .get("/user")
  .then(function(response) {
    // handle success
    console.log(response);
  })
  .catch(function(error) {
    // handle error
    console.log(error);
  })
  .finally(function() {
    // always executed
  });
```

### delete(url[, config])

```js
import Http from "mlz-axios";
const httpIns = new Http("https://www.example.com");
httpIns
  .delete("/user")
  .then(function(response) {
    // handle success
    console.log(response);
  })
  .catch(function(error) {
    // handle error
    console.log(error);
  })
  .finally(function() {
    // always executed
  });
```

### post(url[, data[, config]])

```js
import Http from "mlz-axios";
const httpIns = new Http("https://www.example.com");
httpIns
  .post("/user", {
    firstName: "Fred",
    lastName: "Flintstone"
  })
  .then(function(response) {
    // handle success
    console.log(response);
  })
  .catch(function(error) {
    // handle error
    console.log(error);
  })
  .finally(function() {
    // always executed
  });
```

### put(url[, data[, config]])

```js
import Http from "mlz-axios";
const httpIns = new Http("https://www.example.com");
httpIns
  .put("/user", {
    firstName: "Fred",
    lastName: "Flintstone"
  })
  .then(function(response) {
    // handle success
    console.log(response);
  })
  .catch(function(error) {
    // handle error
    console.log(error);
  })
  .finally(function() {
    // always executed
  });
```

### patch(url[, data[, config]])

```js
import Http from "mlz-axios";
const httpIns = new Http("https://www.example.com");
httpIns
  .patch("/user", {
    firstName: "Fred",
    lastName: "Flintstone"
  })
  .then(function(response) {
    // handle success
    console.log(response);
  })
  .catch(function(error) {
    // handle error
    console.log(error);
  })
  .finally(function() {
    // always executed
  });
```

## 例子
```js
import Http from "@mlz/axios";

const token = localStorage.getItem('authorization')
const AUTHORIZATION_TYPE = 3

// 全局设置token与tokenType
Http.setGlobalTokenConfig({
  authorizationTypeKey: 'Authorization',
  authorizationTokenKey: 'authorization_type',
  authorizationTypeValue: AUTHORIZATION_TYPE,
  authorizationTokenValue: token
})

const httpIns = new Http('https://xxx.aaa.com');
const httpIns2 = new Http('https://xxx.bbb.com');
const httpIns3 = new Http('https://xxx.ccc.com');

const httpIns2Token = localStorage.getItem('authorization_httpIns2')
const AUTHORIZATION_HTTPINS_2_TYPE = 4

// 设置单个实例的token与tokenType
httpIns2.setInstanceTokenConfig({
  authorizationTypeKey: 'Authorization',
  authorizationTokenKey: 'authorization_type',
  authorizationTypeValue: AUTHORIZATION_HTTPINS_2_TYPE,
  authorizationTokenValue: httpIns2Token
})

//对单个实例设置请求拦截器
Http.setReqInterceptor(
  config => {
    console.log(config)
  },
  err => {
    return Promise.reject(err);
  },
  'https://xxx.aaa.com'
);

httpIns
  .post("/xxx/xxx/xxx", {
    id: "xxx",
  })
  .then(res => {})
  .catch(err => {
    console.log(err);
  });

httpIns2
  .get("/xxx/xxx/xxx", { params: { id: 1 } })
  .then(res => {});

httpIns3
  .get("/xxx/xxxx")
  .then(res => {
    if (/^2/.test(res.status)) {
      console.log(res.data);
    } else {
      console.log(res.data.error_code);
    }
  })
  .catch(err => {
    console.log(err);
  });
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/juicecube/mlz-axios/compare/v1.1.2...v1.2.0) (2021-09-24)


### Features

* add mlz-lint ([b93befa](https://github.com/juicecube/mlz-axios/commit/b93befa885432bb9fe4c3266116c5f7424862f48))


### Bug Fixes

* add global header config ([cbe0145](https://github.com/juicecube/mlz-axios/commit/cbe01455dc7602e5f51f9123ab92c5030e4f4930))
* add stand version ([ea56bcc](https://github.com/juicecube/mlz-axios/commit/ea56bcc607bf3228f0c1abd5ac0dc89183cb68ab))
* update axios ([e207302](https://github.com/juicecube/mlz-axios/commit/e207302d05b408fd373deac04da62750325a8ead))
* 更新版本 ([7f3aea0](https://github.com/juicecube/mlz-axios/commit/7f3aea0046ce7c18af5f58f4e0e1030c3e2325e4))


src/mlz_axios.ts
import axios, {
  AxiosRequestConfig,
  CancelTokenSource,
  AxiosPromise,
  AxiosResponse,
  AxiosInstance,
} from 'axios';

type TokenConfig = {
  authorizationTypeKey:string;
  authorizationTokenKey:string;
  authorizationTypeValue:number;
  authorizationTokenValue:string;
}

interface ReqInterceptor {
  resolve:(config:AxiosRequestConfig) => AxiosRequestConfig;
  reject:(error:any) => void;
}

interface ResInterceptor {
  resolve:(res:AxiosResponse) => AxiosResponse;
  reject:(error) => any;
}

const DEFAULT_CONFIG:AxiosRequestConfig = {
  timeout: 0,
  withCredentials: true,
  validateStatus: (status) => status >= 200 && status < 599,
};

export class Http {
  private isSetInstancesToken:boolean = false;

  public axiosIns:AxiosInstance;
  public source:CancelTokenSource = axios.CancelToken.source();
  public instanceTokenConfig:TokenConfig = {
    authorizationTypeKey: 'Authorization',
    authorizationTokenKey: 'authorization_type',
    authorizationTypeValue: 0,
    authorizationTokenValue: '',
  }

  static globalTokenConfig:TokenConfig = {
    authorizationTypeKey: 'Authorization',
    authorizationTokenKey: 'authorization_type',
    authorizationTypeValue: 0,
    authorizationTokenValue: '',
  }

  static globalHeaders = {};

  static INSTANCES:{ [key:string]:AxiosInstance } = {};
  static INSTANCES_REQUEST_INTERCEPTORS:{ [key:string]:ReqInterceptor } = {};
  static INSTANCES_RESPONSE_INTERCEPTORS:{ [key:string]:ResInterceptor } = {};
  static GLOBAL_REQUEST_INTERCEPTORS:ReqInterceptor = null;
  static GLOBAL_RESPONSE_INTERCEPTORS:ResInterceptor = null;

  constructor(baseUrl:string, config:AxiosRequestConfig = {}) {
    const axiosIns = axios.create({
      baseURL: baseUrl,
      ...DEFAULT_CONFIG,
      ...config,
    });
    Http.INSTANCES[baseUrl] = axiosIns;
    this.axiosIns = axiosIns;

    // 设置拦截器
    const requestInterceptor = Http.INSTANCES_REQUEST_INTERCEPTORS[baseUrl] || Http.GLOBAL_REQUEST_INTERCEPTORS;
    const responseInterceptor = Http.INSTANCES_RESPONSE_INTERCEPTORS[baseUrl] || Http.GLOBAL_RESPONSE_INTERCEPTORS;
    if (requestInterceptor) {
      this.axiosIns.interceptors.request.use(requestInterceptor.resolve, requestInterceptor.reject);
    }
    if (responseInterceptor) {
      this.axiosIns.interceptors.response.use(responseInterceptor.resolve, responseInterceptor.reject);
    }
  }

  // 设置单个实例的token与tokenType
  public setInstanceTokenConfig( instanceTokenConfig:Partial<TokenConfig>) {
    this.isSetInstancesToken = true;
    this.instanceTokenConfig = {
      ...this.instanceTokenConfig,
      ...instanceTokenConfig,
    };
  }

  private request(opt:AxiosRequestConfig) {

    const _opt = Object.assign({}, opt);
    if (this.isSetInstancesToken) {
      const { authorizationTypeValue, authorizationTokenValue, authorizationTypeKey, authorizationTokenKey } = this.instanceTokenConfig;
      if (authorizationTypeKey && authorizationTypeValue && authorizationTokenKey && authorizationTokenValue){
        _opt.headers = {
          [authorizationTypeKey]: authorizationTypeValue,
          [authorizationTokenKey]: authorizationTokenValue,
          ..._opt.headers,
        };
      }
    } else {
      const { authorizationTypeValue, authorizationTokenValue, authorizationTypeKey, authorizationTokenKey } = Http.globalTokenConfig;
      if (authorizationTypeKey && authorizationTypeValue && authorizationTokenKey && authorizationTokenValue) {
        _opt.headers = {
          [authorizationTypeKey]: authorizationTypeValue,
          [authorizationTokenKey]: authorizationTokenValue,
          ..._opt.headers,
        };
      }
      if (Http.globalHeaders) {
        console.log(_opt.headers);
        _opt.headers = { ..._opt.headers, ...Http.globalHeaders };
      }
    }
    _opt.cancelToken = this.source.token;
    return this.axiosIns.request(_opt);
  }

  public abort() {
    this.source.cancel('API abort.');
  }
  public get(url:string, configs?:AxiosRequestConfig):AxiosPromise {
    return this.request({
      method: 'get',
      url,
      ...configs,
    });
  }
  public post(url:string, data?:any, configs?:AxiosRequestConfig):AxiosPromise {
    return this.request({
      method: 'post',
      url,
      data,
      ...configs,
    });
  }
  public put(url:string, data?:any, configs?:AxiosRequestConfig):AxiosPromise {
    return this.request({
      method: 'put',
      url,
      data,
      ...configs,
    });
  }
  public patch(url:string, data?:any, configs?:AxiosRequestConfig):AxiosPromise {
    return this.request({
      method: 'patch',
      url,
      data,
      ...configs,
    });
  }
  public delete(url:string, configs?:AxiosRequestConfig):AxiosPromise {
    return this.request({
      method: 'delete',
      url,
      ...configs,
    });
  }

  static setGlobalHeaders(data:{[key:string]:any}) {
    Http.globalHeaders = data;
  }

  // 全局设置token与tokenType
  static setGlobalTokenConfig(globalTokenConfig:Partial<TokenConfig>) {
    Http.globalTokenConfig = {
      ...Http.globalTokenConfig,
      ...globalTokenConfig,
    };
  }

  // 获取相对应的实例
  static getInstances(key:string):AxiosInstance {
    return Http.INSTANCES[key];
  }

  // 设置请求拦截器
  static setReqInterceptor(
    resolve?:(config:AxiosRequestConfig) => AxiosRequestConfig,
    reject?:(error:any) => void,
    url?:string,
  ) {
    if (url) {
      // 单个实例
      Http.INSTANCES_REQUEST_INTERCEPTORS[url] = {
        resolve,
        reject,
      };
    } else {
      // 全局
      Http.GLOBAL_REQUEST_INTERCEPTORS = {
        resolve,
        reject,
      };
    }
  }

  // 设置响应拦截器
  static setResInterceptor(
    resolve?:(res:AxiosResponse) => AxiosResponse,
    reject?:(error) => any,
    url?:string,
  ) {
    if (url) {
      // 实例
      Http.INSTANCES_RESPONSE_INTERCEPTORS[url] = {
        resolve,
        reject,
      };
    } else {
      // 全局
      Http.GLOBAL_RESPONSE_INTERCEPTORS = {
        resolve,
        reject,
      };
    }
  }
}
